README.md
# PetroCloud Utils (pcutils)

---

[![npm version](https://badge.fury.io/js/pcutils.svg)](https://badge.fury.io/js/pcutils)
[![Build Status](https://travis-ci.org/PetroCloud/pcutils.svg?branch=master)](https://travis-ci.org/PetroCloud/pcutils)
[![Coverage Status](https://coveralls.io/repos/github/PetroCloud/pcutils/badge.svg)](https://coveralls.io/github/PetroCloud/pcutils)

---

Some utility functions we use for several projects

## Change Log

See the [CHANGELOG.md](CHANGELOG.md) for information on what's new and what changed.

## Usage

### Function `httpRequest(options)`

Executes an HTTP request.
- `options` request options. See [`request` options documentation](https://github.com/request/request#requestoptions-callback)
- returns a Bluebird promise that is resolved with the `request` response 
object, unless `options.returnBody` is set to `true`, in that case, it resolves
the promise with the body content.

In the case where the request returns an HTTP error (anything from 400 to 599 
HTTP status code), it will reject the promise with an error object with 
`err.message` = `IncomingMessage.statusMessage` as it's message, and 
`err.statusCode` = `IncomingMessage.statusCode`. If `options.returnBody` is set 
to `false`, the `err` object returned is augmented with the response object.

### Function `postJSONObject(url, objToSend, timeout, returnBody=true)` 
#### Aliased to `postJSON`, too.

Sends a JSON object to and endpoint as an HTTP POST
- `url` URL string of the endpoint
- `objToSend` object to send
- `timeout` integer containing the number of milliseconds to
 wait for a server to send response headers (and start the response body)
 before aborting the request. Note that if the underlying TCP connection
 cannot be established, the OS-wide TCP connection timeout will overrule the
 timeout option (the default in Linux can be anywhere from 20-120 seconds).
- `returnBody` boolean that defaults to `true`. Instructs the function to return
  body if true, or the response object response if false.
- returns a Bluebird promise that is resolved with the `request` response 
body, unless `returnBody` is set to `false`, in that case, it resolves
the promise with the response object.

In the case where the request returns an HTTP error (anything from 400 to 599 
HTTP status code), it will reject the promise with an error object with 
`err.message` = `IncomingMessage.statusMessage` as it's message, and 
`err.statusCode` = `IncomingMessage.statusCode`.
If `returnBody` is set to `false`, the `err` object returned is augmented with 
the response object.

### Function `getJSON(url, timeout)`

Get a JSON object from an endpoint as an HTTP GET
- `url` URL string of the endpoint
- `timeout` integer containing the number of milliseconds to
 wait for a server to send response headers (and start the response body)
 before aborting the request. Note that if the underlying TCP connection
 cannot be established, the OS-wide TCP connection timeout will overrule the
 timeout option (the default in Linux can be anywhere from 20-120 seconds).
- `returnBody` boolean that defaults to `true`. Instructs the function to return
  body if true, or the response object response if false.
- returns a Bluebird promise that is resolved with the `request` response 
body, unless `returnBody` is set to `false`, in that case, it resolves
the promise with the response object.

In the case where the request returns an HTTP error (anything from 400 to 599 
HTTP status code), it will reject the promise with an error object with 
`err.message` = `IncomingMessage.statusMessage` as it's message, and 
`err.statusCode` = `IncomingMessage.statusCode`.
If `returnBody` is set to `false`, the `err` object returned is augmented with 
the response object.

### `promisifier(function)`

Takes a function as a parameter and returns a promise that resolves when the
function returns a value or rejects if an exception ocurrs
- `function` is the function to promisify
- returns a Bluebird promise with the result of the call to the function or an
error if one is thrown

### `objectToString(object, showHidden, depth)`

Takes an object or an array and returns the 'deep' representation of it.

- `object` is the object subject to be converted
- `showHidden` (defaults to false) makes the function show hidden attributes
- `depth` (defaults to null, that is, infinite depth) tells how deep to go in
the nesting chain, when converting the object as a string.
- returns a String that resembles the object, based on the parameters passed

## License

**[MIT](./LICENSE)**
&copy; 2014-2016
[Petrocloud](http://petrocloud.com)

This module is free and open-source under the MIT License.


CHANGELOG.md
# PetroCloud Utils (pcutils) CHANGELOG

## [Unreleased] yet

## [2.0.0] - 2018-01-01
### Changed
- Removes all unused functions `existsIn`, `promisifier`, `objectToString` [@luislobo]
- Updated dependencies [@luislobo]
- Supports async/await. From now on, only compatible with Node 8+ [@luislobo]
- Replaced jshint by eslint [@luislobo]

## [1.0.21] - 2017-01-04
### Changed
- `getJSON`, `POSTJSON` and `httpRequest` now rejects when it gets a statusCode 
 between 400 and 599 [@luislobo]

## [1.0.20] - 2016-11-07
### Changed
- Links to Coveralls badges [@luislobo]

## [1.0.19] - 2016-11-07
### Added
- Posting code coverage to Coveralls [@luislobo]

## [1.0.18] - 2016-11-07
### Changed
- Upgraded modules [@luislobo]
- Updated `README.md` with documentation about `promisifier` and `objectToString` functions [@luislobo]

## [1.0.17] - 2016-11-07
### Added
- `promisifier` and `objectToString` functions [@luislobo]

## [1.0.16] - 2016-03-16
### Added
- istanbul on the test script [@luislobo]
- `CHANGELOG.md` [@luislobo]

## [1.0.15]  - 2016-05-16
### Added
- returnBody parameter to getJSON and postJSON [@luislobo]

### Changed
- Upgraded packages [@luislobo]

## [1.0.14] - 2016-03-14
### Added
- Travis badge [@luislobo]

### Fixed
- Fix unit tests [@luislobo]

## [1.0.13] - 2016-03-14
### Added
- Badges to `README.md` [@luislobo]

### Changed
- Travis config file [@luislobo]

## [1.0.12] - 2016-03-14
### Changed
- Made getJSON and postJSON return the response body instead of the request object [@luislobo]
- Updated `README.md` [@luislobo]

## [1.0.11] - 2016-03-13
---
- general cleanup [@luislobo]
- replaced Q with Bluebird [@luislobo]
- Added unit tests and coverage [@luislobo]

[@luislobo]: https://github.com/luislobo/
[Unreleased]: https://github.com/PetroCloud/pcutils/compare/v2.0.0...HEAD
[2.0.0]: https://github.com/PetroCloud/pcutils/compare/v1.0.21...v2.0.0
[1.0.21]: https://github.com/PetroCloud/pcutils/compare/v1.0.20...v1.0.21
[1.0.20]: https://github.com/PetroCloud/pcutils/compare/v1.0.19...v1.0.20
[1.0.19]: https://github.com/PetroCloud/pcutils/compare/v1.0.18...v1.0.19
[1.0.18]: https://github.com/PetroCloud/pcutils/compare/v1.0.17...v1.0.18
[1.0.17]: https://github.com/PetroCloud/pcutils/compare/v1.0.16...v1.0.17
[1.0.16]: https://github.com/PetroCloud/pcutils/compare/v1.0.15...v1.0.16
[1.0.15]: https://github.com/PetroCloud/pcutils/compare/v1.0.14...v1.0.15
[1.0.14]: https://github.com/PetroCloud/pcutils/compare/v1.0.13...v1.0.14
[1.0.13]: https://github.com/PetroCloud/pcutils/compare/v1.0.12...v1.0.13
[1.0.12]: https://github.com/PetroCloud/pcutils/compare/v1.0.11...v1.0.12
[1.0.11]: https://github.com/PetroCloud/pcutils/compare/90d32912b84b8921ad759af5ff9410cb80472822...v1.0.11


test/pcutil.spec.js
const pu = require('../lib/utils');
const _ = require('lodash');
const chai = require('chai');
chai.should();
const chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const http = require('http');
const url = require('url');

function createEchoServer(port, delay){
  delay = delay || 0;
  const server = http.createServer((req, resp) =>{
    const parsedUrl = url.parse(req.url, true); // true to get query as object
    const queryAsObject = parsedUrl.query;

    let dataToSendBack = '';
    req.on('data', (chunk) =>{
      dataToSendBack += chunk;
    });
    req.on('end', () =>{
      if (req.method === 'GET') {
        dataToSendBack = {
          some: 'sample',
          is: 'here'
        };
      }
      _.delay(() =>{
        if (queryAsObject.reject) {
          resp.statusCode = 500;
          resp.write('Server error');
          resp.end();
        } else {
          resp.writeHead(200, {'content-type': 'application/json'});
          resp.write(JSON.stringify({
            url: req.url,
            method: req.method,
            headers: req.headers,
            result: dataToSendBack
          }));
          resp.end();
        }
      }, delay);
    });
  });
  server.port = port;
  server.url = `http://localhost:${port}`;
  server.protocol = 'http';
  return server;
}

describe('httpRequest', () =>{

  const server = createEchoServer(6789);
  const options = {
    uri: 'http://127.0.0.1:6789'
  };

  before((done) =>{
    server.listen(server.port, done);
  });

  after((done) =>{
    server.close(done);
  });

  it('should throw a reject when the server is not available',
    () => pu.httpRequest({
      uri: 'http://127.0.0.1:1234/',
      returnBody: true
    }).should.be.rejectedWith(Error));

  it('should throw a reject when the server is not available even if returnBody is false',
    () => pu.httpRequest({
      uri: 'http://127.0.0.1:1234/',
      returnBody: false
    }).should.be.rejectedWith(Error));

  it('should throw a reject when the server returns an error',
    () => pu.httpRequest({
      uri: 'http://127.0.0.1:6789/?reject=1',
      returnBody: true
    }).should.be.rejectedWith(Error));

  it('should throw a reject when the server returns an error even if returnBody is false',
    () => pu.httpRequest({
      uri: 'http://127.0.0.1:6789/?reject=1',
      returnBody: false
    }).should.be.rejectedWith(Error));

  it('should GET a request returning body', () => pu.httpRequest(_.merge({
    returnBody: true
  }, options))
    .then((res) =>{
      res = JSON.parse(res);
      res.should.have.property('result');
    }));

  it('should GET a request with a header', () => pu.httpRequest(_.merge({
    returnBody: true,
    header: {
      'Content-Type': 'application/json'
    }
  }, options))
    .then((res) =>{
      res = JSON.parse(res);
      res.should.have.property('result');
    }));

  it('should GET a request and return a response object instead of body',
    () => pu.httpRequest(options)
      .then((res) =>{
        res.should.have.property('request');
        res.should.have.property('statusCode', 200);
        res.request.should.have.property('href', 'http://127.0.0.1:6789/');
        res.should.have.property('body');
        const respObj = JSON.parse(res.body);
        respObj.should.have.property('result');
        respObj.result.should.have.property('some');
      }));

  it('should POST a request', () => pu.httpRequest(_.merge({
    body: 'This is a POST body',
    method: 'POST'
  }, options))
    .then((res) =>{
      res.should.have.property('body');
      const respObj = JSON.parse(res.body);
      respObj.should.have.property('result');
      respObj.result.should.equal('This is a POST body');
    }));

  it('should POST a request as JSON', () => pu.httpRequest(_.merge({
    json: true,
    body: {
      some: 'sample',
      is: 'here'
    },
    method: 'POST'
  }, options))
    .then((res) =>{
      res.should.have.property('headers');
      res.headers.should.have.property('content-type', 'application/json');
      res.should.have.property('body');
      const respObj = res.body;
      respObj.should.have.property('result');
      const result = JSON.parse(respObj.result);
      result.should.have.property('some', 'sample');
      result.should.have.property('is', 'here');
    }));

});


describe('postJSONObject', () =>{

  const server = createEchoServer(6789, 100);

  before((done) =>{
    server.listen(server.port, done);
  });

  after((done) =>{
    server.close(done);
  });

  it('should throw a reject when the server is not available',
    () => pu.postJSONObject('http://127.0.0.1:1234/').should.be.rejectedWith(Error));

  it('should throw a reject when the server returns an error',
    () => pu.postJSONObject('http://127.0.0.1:6789/?reject=1').should.be.rejectedWith(Error));


  it('should POST a request as JSON', () => pu.postJSONObject('http://127.0.0.1:6789/',
    {
      some: 'sample',
      is: 'here'
    })
    .then((res) =>{
      res.should.have.property('headers');
      res.headers.should.have.property('content-type', 'application/json');
      res.should.have.property('result');
      const result = JSON.parse(res.result);
      result.should.have.property('some', 'sample');
      result.should.have.property('is', 'here');
    }));

  it('should POST a request as JSON and return as Body for wrong returnBody param', () =>
    pu.postJSONObject('http://127.0.0.1:6789/',
      {
        some: 'sample',
        is: 'here'
      },null, 1) // <= this should be boolean but as a safeguard, it returns Body when it's not boolean
      .then((res) =>{
        res.should.have.property('headers');
        res.headers.should.have.property('content-type', 'application/json');
        res.should.have.property('result');
        const result = JSON.parse(res.result);
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      }));

  it('should POST a request as JSON using alias', () =>
    pu.postJSON('http://127.0.0.1:6789/',
      {
        some: 'sample',
        is: 'here'
      })
      .then((res) =>{
        res.should.have.property('headers');
        res.headers.should.have.property('content-type', 'application/json');
        res.should.have.property('result');
        const result = JSON.parse(res.result);
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      }));

  it('should POST a request as JSON but returning a request', () =>

    // last param is returnBody = false
    pu.postJSONObject('http://127.0.0.1:6789/',
      {
        some: 'sample',
        is: 'here'
      }, null, false)
      .then((res) =>{
        res.should.have.property('request');
        res.should.have.property('statusCode', 200);
        res.should.have.property('connection');
        res.connection.should.have.property('bytesRead');
        res.connection.should.have.property('bytesWritten');
        res.request.should.have.property('href', 'http://127.0.0.1:6789/');
        res.should.have.property('body');
        res.body.should.have.property('result');
        const result = JSON.parse(res.body.result);
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      })
  );

  it('should return timeout sending a POST request as JSON with 50ms',
    () => pu.postJSONObject('http://127.0.0.1:6789/',
      {
        some: 'sample',
        is: 'here'
      }, 50)
      .should.be.rejectedWith(Error));

  it('should POST a request as JSON with a 150ms timeout', () => pu.postJSONObject('http://127.0.0.1:6789/',
    {
      some: 'sample',
      is: 'here'
    }, 150)
    .then((res) =>{
      res.should.have.property('headers');
      res.headers.should.have.property('content-type', 'application/json');
      res.should.have.property('result');
      const result = JSON.parse(res.result);
      result.should.have.property('some', 'sample');
      result.should.have.property('is', 'here');
    }));

});


describe('getJSON', () =>{

  const server = createEchoServer(6789, 100);

  before((done) =>{
    server.listen(server.port, done);
  });

  after((done) =>{
    server.close(done);
  });

  it('should throw a reject when the server is not available',
    () => pu.getJSON('http://127.0.0.1:1234/').should.be.rejectedWith(Error));

  it('should throw a reject when the server returns an error',
    () => pu.getJSON('http://127.0.0.1:6789/?reject=1').should.be.rejectedWith(Error));

  it('should throw a reject when the port is > 65535',
    () => pu.getJSON('http://127.0.0.1:66666').should.be.rejectedWith(Error));

  it('should GET a response as JSON',
    () => pu.getJSON('http://127.0.0.1:6789/')
      .then((res) =>{
        res.should.have.property('result');
        const result = res.result;
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      }));

  it('should GET a response as JSON and return as Body for wrong returnBody param',
    () => pu.getJSON('http://127.0.0.1:6789/', null, 1) // <== 1 should be a boolean
      .then((res) =>{
        res.should.have.property('result');
        const result = res.result;
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      }));

  it('should GET a response as JSON, returning a response object',
    () =>
      // last param is returnBody = false
      pu.getJSON('http://127.0.0.1:6789/', null, false)
        .then((resp) =>{
          resp.should.have.property('request');
          resp.should.have.property('statusCode', 200);
          resp.should.have.property('connection');
          resp.connection.should.have.property('bytesRead');
          resp.connection.should.have.property('bytesWritten');
          resp.request.should.have.property('href', 'http://127.0.0.1:6789/');
          resp.should.have.property('body');
          resp.body.should.have.property('result');
          const result = resp.body.result;
          result.should.have.property('some', 'sample');
          result.should.have.property('is', 'here');
        })
  );

  it('should return timeout sending a GET request as JSON with 50ms',
    () => pu.getJSON('http://127.0.0.1:6789/', 50).should.be.rejectedWith(Error));

  it('should GET a request as JSON with a 150ms timeout',
    () => pu.getJSON('http://127.0.0.1:6789/', 150)
      .then((res) =>{
        res.should.have.property('result');
        const result = res.result;
        result.should.have.property('some', 'sample');
        result.should.have.property('is', 'here');
      }));

});


Gruntfile.js
/* global module:false */
module.exports = function (grunt) {

  // Project configuration.
  grunt.initConfig({
    // Task configuration.
    eslint: {
      options: {
        format: 'codeframe',
        fix: true
      },
      gruntfile: {
        src: 'Gruntfile.js'
      },
      lib_test: {
        src: ['lib/*.js', 'test/*.spec.js']
      }
    },
    mochaTest: {
      test: {
        options: {
          reporter: 'spec',
          timeout: 5000
        },
        src: ['test/*.spec.js']
      }
    },
    mocha_istanbul: {
      coverage: {
        options: {
          reporter: 'spec',
          timeout: 5000
        },
        src: ['test/*.spec.js']
      }
    },
    watch: {
      gruntfile: {
        files: '<%= jshint.gruntfile.src %>',
        tasks: ['eslint:gruntfile']
      },
      lib_test: {
        files: '<%= jshint.lib_test.src %>',
        tasks: ['eslint:lib_test', 'mochaTest']
      },
      default: {
        files: 'lib/*.js',
        tasks: ['eslint']
      }
    }
  });

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-mocha-test');
  grunt.loadNpmTasks('grunt-mocha-istanbul');
  grunt.loadNpmTasks('gruntify-eslint');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-coveralls');

  // testing
  grunt.registerTask('test', ['mochaTest']);
  grunt.registerTask('coverage', ['mocha_istanbul:coverage']);
  // Default task.
  grunt.registerTask('default', ['eslint', 'test']);


};


lib/pc3utils.js
var Q = require("q");
var request = Q.denodeify(require("request"));

function condense(options) {
  return request(options)
    .spread(function(response, body) {
      var code = response.statusCode;
      if (code === 200) {
        return Q(body);
      } else {
        throw new Error(JSON.stringify({statusCode: response.statusCode, body: body}));
      }
    });
}

function success(jsonObject) {
  var _ = require("lodash");
  return _.merge({
    response: "ok"
  }, jsonObject);
}

function fail(errCode, errMessage, jsonObject) {
  var _ = require("lodash");
  return _.merge({
    response: "fail",
    code: errCode,
    message: errMessage
  }, jsonObject);
}

function postJSONObject(url, JSONObject, timeout) {
  var options = {
    uri: url,
    json: true,
    body: JSONObject,
    method: "post"
  };
  if (timeout) {
    options.timeout = timeout;
  }
  return condense(options);
}

function postXML(url, xml) {
  var options = {
    uri: url,
    body: xml,
    headers: {
      'Content-Type': 'text/xml'
    },
    method: "post"
  };
  return condense(options);
}

function get(url) {
  var options = {
    uri: url,
    method: "get"
  };
  return condense(options);
}

function getJSON(url) {
  var options = {
    uri: url,
    json: true,
    header: {
      'Content-Type': 'application/json'
    },
    method: "get"
  };
  return condense(options);
}

function existsIn(sourceArray, itemsToMatch) {
  return sourceArray.some(function(sourceElement) {
    return itemsToMatch.some(function(matchElement) {
      return matchElement === sourceElement;
    });
  });
}

// Functions which will be available to external callers
module.exports.success = success;
module.exports.fail = fail;
module.exports.postJSONObject = postJSONObject;
module.exports.postXML = postXML;
module.exports.get = get;
module.exports.getJSON = getJSON;
module.exports.existsIn = existsIn;


lib/utils.js
/* global Promise */
const Promise = require('bluebird');
const request = require('request');
const _ = require('lodash');

/**
 * Executes an HTTP request.
 *
 * @param {Object} options request options object. See
 * https://github.com/request/request#requestoptions-callback
 * @returns {Promise} returns a promise where the body is the data of the
 * request
 */
async function httpRequest(options) {
  return new Promise((resolve, reject) => {
    request(options, (reqErr, response, body) => {
      if (reqErr) {
        return reject(reqErr);
      }
      if (response.statusCode >= 400 && response.statusCode <= 599) { // eslint-disable-line no-magic-numbers
        let err = new Error(response.statusMessage);
        err.statusMessage = response.statusMessage;
        err.statusCode = response.statusCode;
        if (!options.returnBody) {
          err = _.merge(err, response);
        }
        return reject(err);
      }
      if (options.returnBody) {
        return resolve(body);
      }
      return resolve(response);
    });
  });
}

/**
 * Sends a JSON object to and endpoint as an HTTP POST
 *
 * @alias postJSON
 * @param {string} url URL of the endpoint
 * @param {Object} objToSend Object to send
 * @param {number} timeout Integer containing the number of milliseconds to
 * wait for a server to send response headers (and start the response body)
 * before aborting the request. Note that if the underlying TCP connection
 * cannot be established, the OS-wide TCP connection timeout will overrule the
 * timeout option (the default in Linux can be anywhere from 20-120 seconds).
 * @param {boolean} returnBody defaults to `true`. Instructs the function to return
 * body if true, or the response object response if false.
 * @returns {Promise} returns a promise where the body is the data of the
 * response if the param returnBody evaluates to true, or the response object
 * itself, if the returnBody parameter evaluates to false (like when not passed
 * at all)
 */
async function postJSON(url, objToSend, timeout, returnBody = true) {
  if (!_.isBoolean(returnBody)) {
    returnBody = true;
  }
  const options = {
    uri: url,
    json: true,
    body: objToSend,
    header: {
      'Content-Type': 'application/json'
    },
    method: 'POST',
    returnBody
  };
  if (_.isNumber(timeout)) {
    options.timeout = timeout;
  }
  return httpRequest(options);
}

/**
 * Get a JSON object from an endpoint as an HTTP GET
 * @param {string} url URL of the endpoint
 * @param {number} timeout Integer containing the number of milliseconds to
 * wait for a server to send response headers (and start the response body)
 * before aborting the request. Note that if the underlying TCP connection
 * cannot be established, the OS-wide TCP connection timeout will overrule the
 * timeout option (the default in Linux can be anywhere from 20-120 seconds).
 * @param {boolean} returnBody defaults to `true`. Instructs the function to return
 * body if true, or request if false.
 * @returns {Promise} returns a promise where the body is the data of the
 * response if the param returnBody evaluates to true, or the response object
 * itself, if the returnBody parameter evaluates to false (like when not passed
 * at all)
 */
async function getJSON(url, timeout, returnBody = true) {
  if (!_.isBoolean(returnBody)) {
    returnBody = true;
  }
  const options = {
    uri: url,
    json: true,
    header: {
      'Content-Type': 'application/json'
    },
    method: 'GET',
    returnBody
  };
  if (_.isNumber(timeout)) {
    options.timeout = timeout;
  }
  return httpRequest(options);
}

// Functions which will be available to external callers
module.exports.httpRequest = httpRequest;
module.exports.postJSONObject = postJSON;
module.exports.postJSON = postJSON;
module.exports.getJSON = getJSON;
